### PR TITLE
GrafanaData: Add now parameter to dateMath.parse to subsitute now string

### LINKED ISSUE #96547: Dashboard: "last 24h view" is constantly flickering
### What happened?

It seems since the update to grafana-enterprise 11.3 (2024-10-24 00:40:14) the any dashboard with the 'last 24h' view is constantly flickering for the x-axis.
One fraction of a second it shows the date and hour then, for the remainder of the second, just the hour.

### What did you expect to happen?

The dashboard x-axis should not flicker and show the past 24 hours.

### Did this work before?

I cannot be 100% sure but it worked with grafana-enterprise 11.2 and prior.

### How do we reproduce it?

1. Update to grafana-enterprise 11.3 (most likely without following the additional instructions, if any) 
`/usr/bin/zypper --quiet --non-interactive --xmlout update --type package --auto-agree-with-licenses --no-recommends`
2. Go to any dashboard and selevt the 'last 24h' filter
3. move the cursor away from the panels, as a cursor inside the panels stops the flickering (at least sometimes)

### Is the bug inside a dashboard panel?

_No response_

### Environment (with versions)?

Grafana: enterprise 11.3
OS: openSUSE Leap 15.6
Browser: Mozilla FF (132.0.2 64-bit) and Chrome (126.0.6478.127)

### Grafana platform?

A package manager (APT, YUM, BREW, etc.)

### Datasource(s)?

_No response_

### PR DESCRIPTION
**Pull request by oscarkilhed**
**What is this feature?**

This adds a `now` parameter to dateMath.parse that will be used to substitute `"now"`

**Why do we need this feature?**

If dateMath.parse is called two times in a row we're not guaranteed to get the same value because the clock might tick over to the next millisecond. This is problematic when calculating time ranges.

The problem:
Time is 2024-01-02 00:00:00:000
Calculate from (now-24h) result: 2024-01-01 00:00:00:000
Clock tics over to 2024-01-02 00:00:00:001
Calculate to (now) result: 2024-01-02 00:00:00:001

The diff between from and too is 86400001 (24h and 1ms). This creates issues down the line where timerange > 24h evaluates to true.

Fixes #96547

**Note**

To fix this https://github.com/grafana/scenes/pull/981 also needs to be merged and the scenes dependency updated in Grafana.

### CODE REVIEW

File: packages/grafana-data/src/datetime/datemath.ts [hunk 1]
```diff
@@ -40,12 +41,14 @@ export function isMathString(text: string | DateTime | Date): boolean {
...
  * @param timezone Only string 'utc' is acceptable here, for anything else, local timezone is used.
+ * @param now DateTimeInput to use as now. Useful when parsing multiple values and now needs to be the same. Without this, comparing results from subsequent parses is not guaranteed to be deterministic.
  */
 export function parse(
   text?: string | DateTime | Date | null,
   roundUp?: boolean,
   timezone?: TimeZone,
-  fiscalYearStartMonth?: number
+  fiscalYearStartMonth?: number,
+  now?: DateTimeInput
```
Comment (aocenas) [inline packages/grafana-data/src/datetime/datemath.ts:51]: I feel like at this point, the arg list is pretty horrible, considering it's 5 params all optional :D. At what point should we create a new function, with an options object? @ashharrison90 @joshhunt

Comment (joshhunt) [reply packages/grafana-data/src/datetime/datemath.ts:51]: Yes, it’s quite unwieldy. Why is the text optional? 😂 

I think an overload is better on the balance of all things.

Comment (joshhunt) [reply packages/grafana-data/src/datetime/datemath.ts:51]: The text to parse should be mandatory first parameter with optional options object.

Comment (oscarkilhed) [reply packages/grafana-data/src/datetime/datemath.ts]: @aocenas @joshhunt I hear you. I've created a new function and marked this one as deprecated. Let me know what you think.

File: packages/grafana-data/src/datetime/datemath.ts [hunk 2]
```diff
@@ -50,14 +52,41 @@ export function parse(
...
+
+export interface ConversionOptions {
+  /**
+   * Set the time to endOf time unit, otherwise to startOf time unit.
+   */
+  roundUp?: boolean;
+  /**
+   * Only string 'utc' is acceptable here, for anything else, local timezone is used.
+   */
+  timezone?: TimeZone;
```
Comment (Sergej-Vlasov) [inline packages/grafana-data/src/datetime/datemath.ts:66]: deprecated import of `TimeZone` suggests to use newer from `import { TimeZone } from '@grafana/schema';
`

File: packages/grafana-data/src/datetime/datemath.ts [hunk 3]
```diff
@@ -50,14 +52,41 @@ export function parse(
...
+   * DateTimeInput to use as now. Useful when parsing multiple values and now needs to be the same. Without this, comparing results from subsequent parses is not guaranteed to be deterministic.
+   */
+  now?: DateTimeInput;
+}
 
-  if (typeof text !== 'string') {
-    if (isDateTime(text)) {
-      return text;
+/**
+ * @param dateTimeRep A DateTime object, a Date object or a string represenation of a specific time.
```
Comment (Sergej-Vlasov) [inline packages/grafana-data/src/datetime/datemath.ts:78]: typo in `represenation` -> `representation`

**Review by Sergej-Vlasov — APPROVED** [review]
LGTM! Tested the solution and don't see the flicker anymore.
Few minor admin comments:

**Comment by ivanortegaalba** [comment]
@aocenas @joshhunt can you help us to get this merged? We have an escalation waiting for this fix :) Thank you!

**Review by ivanortegaalba — APPROVED** [review]
It is good to me the new function. How costly is to replace it already in every place where it is used in Grafana? Can we have a follow-up PR to replace it?

**Review by aocenas — APPROVED** [review]

**Comment by aocenas** [comment]
@ivanortegaalba sorry for the delay, looks good from my end.

**Comment by Sergej-Vlasov** [comment]
@oscarkilhed I believe this should be backported to `11.3` not `10.3` right?

**Comment by oscarkilhed** [comment]
@Sergej-Vlasov Yes. Thank you

**Pull request merged**